tools/wptrunner/wptrunner/manifestinclude.py

```diff
@@ -81,74 +81,52 @@ def _get_components(self, url):
-                                continue
-                        urls.append(url)
-                        found = True
-                    if found:
-                        break
-        else:
-            urls = [url]
-
-        assert direction in ("include", "exclude")
-
-        for url in urls:
-            components = self._get_components(url)
-
-            node = self
-            while components:
-                component = components.pop()
-                if component not in node.child_map:
-                    new_node = IncludeManifest(DataNode(component))
-                    node.append(new_node)
-                    new_node.set("skip", node.get("skip", {}))
-
-                node = node.child_map[component]
-
-            skip = False if direction == "include" else True
-            node.set("skip", str(skip))
-
-    def add_include(self, test_manifests, url_prefix):
-        """Add a rule indicating that tests under a url path
-        should be included in test runs
-
-        :param url_prefix: The url prefix to include
-        """
-        return self._add_rule(test_manifests, url_prefix, "include")
-
-    def add_exclude(self, test_manifests, url_prefix):
-        """Add a rule indicating that tests under a url path
-        should be excluded from test runs
 
-        :param url_prefix: The url prefix to exclude
-        """
-        return self._add_rule(test_manifests, url_prefix, "exclude")
+def resolve_pattern(test_manifests, pattern):
+    maybe_path = os.path.join(os.path.abspath(os.curdir), pattern)
+    rest, last = os.path.split(maybe_path)
+    fragment = query = None
+    if "#" in last:
+        last, fragment = last.rsplit("#", 1)
+    if "?" in last:
+        last, query = last.rsplit("?", 1)
+
+    maybe_path = os.path.join(rest, last)
+    paths = glob.glob(maybe_path)
+
+    urls = []
+    if paths:
+        for path in paths:
+            for manifest, data in test_manifests.items():
+                found = False
+                rel_path = os.path.relpath(path, data["tests_path"])
+                iterator = manifest.iterpath if os.path.isfile(path) else manifest.iterdir
+                for test in iterator(rel_path):
+                    if not hasattr(test, "url"):
+                        continue
+                    url = test.url
+                    if query or fragment:
+                        parsed = urlparse(url)
+                        if ((query and query != parsed.query) or
+                            (fragment and fragment != parsed.fragment)):
+                            continue
+                    urls.append(url)
+                    found = True
+                if found:
+                    break
+    else:
+        # Turn this URL prefix into a relative path with system separators, as
+        # `Manifest.iter*` expect.
+        rel_path = os.path.normpath(pattern.replace("/", os.path.sep))
```

Comment:
> Before this change we'd run all the variants, now we don't run anything.

Not sure what you mean. I think `iterpath()` still returns all variants? See `test_filter_resolve_prefix_to_variants` (added).

(`iterdir()`, despite the name, also does works on file paths, so deduplicated the URLs at the end.)

> It seems likely you want to strip off the url base, rather than the path separator.

Good catch, thanks!

Reply:
`/wpt run firefox /url/url-constructor.any.html` runs 4 tests on master, but 0 after this PR is applied.